### DISCUSSION TITLE
Can I convert the char array type to json form using NLOHMANN_DEFINE_TYPE_*?

### DISCUSSION THREAD

**Discussion by DrawingProcess** [nlohmann/json#4102] [Q&A]
Can I convert the array type to json form using NLOHMANN_DEFINE_TYPE_*?

```
typedef struct
{
	short				magic;
	char				msg_type;
	short				crc16;
	short				len;
	char				data[0];
} __attribute__((__packed__)) CdaMsgHdr;
NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE_WITH_DEFAULT(CdaMsgHdr, magic, msg_type, crc16, len);
```

**Comment by nlohmann** [comment]
Not directly. char data[0] is a C flexible/zero-length array member — it has no serializable size by itself (sizeof gives 0 elements), and the NLOHMANN_JSON_TO/FROM macro expansion (nlohmann_json_j[#v1] = nlohmann_json_t.v1) would try to treat it as a fixed-size array type, which doesn't apply here since its actual length is determined at runtime by the surrounding buffer. That's presumably why it was left out of your NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE_WITH_DEFAULT(CdaMsgHdr, magic, msg_type, crc16, len) call. You'll need to handle the trailing 'data' payload manually (e.g. as a separate std::vector<char>/std::string field with an explicit length, converted via a hand-written to_json/from_json) rather than through the macro.

---
*This reply was drafted by [Claude Code](https://claude.com/claude-code) on behalf of @nlohmann, as part of a review of open discussions.*
